core/main/src/broker/thunder_broker.rs

```diff
@@ -0,0 +1,222 @@
+// Copyright 2023 Comcast Cable Communications Management, LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+use super::{
+    endpoint_broker::{BrokerCallback, BrokerOutput, BrokerRequest, BrokerSender, EndpointBroker},
+    rules_engine::RuleEndpoint,
+};
+use futures_util::{SinkExt, StreamExt};
+use ripple_sdk::{
+    api::gateway::rpc_gateway_api::JsonRpcApiResponse,
+    log::{debug, error, info},
+    tokio::{self, net::TcpStream, sync::mpsc},
+    utils::error::RippleError,
+};
+use serde_json::json;
+use std::{
+    collections::HashMap,
+    sync::{Arc, RwLock},
+    time::Duration,
+};
+use tokio_tungstenite::client_async;
+
+fn extract_tcp_port(url: &str) -> String {
+    let url_split: Vec<&str> = url.split("://").collect();
+    if let Some(domain) = url_split.get(1) {
+        let domain_split: Vec<&str> = domain.split('/').collect();
+        domain_split.first().unwrap().to_string()
+    } else {
+        url.to_owned()
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct ThunderBroker {
+    sender: BrokerSender,
+    subscription_map: Arc<RwLock<HashMap<String, BrokerRequest>>>,
+}
+
+impl ThunderBroker {
+    fn start(endpoint: RuleEndpoint, callback: BrokerCallback) -> Self {
+        let (tx, mut tr) = mpsc::channel(10);
+        let sender = BrokerSender { sender: tx };
+        let subscription_map = Arc::new(RwLock::new(HashMap::new()));
+        let broker = Self {
+            sender,
+            subscription_map,
+        };
+        let broker_c = broker.clone();
+        let callback_for_sender = callback.clone();
+        tokio::spawn(async move {
+            info!("Broker Endpoint url {}", endpoint.url);
+            let url = url::Url::parse(&endpoint.url).unwrap();
+            let port = extract_tcp_port(&endpoint.url);
+            info!("Url host str {}", url.host_str().unwrap());
+            //let tcp_url = url.host_str()
+            let tcp = loop {
+                if let Ok(v) = TcpStream::connect(&port).await {
+                    break v;
+                } else {
+                    error!("Broker Wait for a sec and retry {}", port);
```

Comment:
I wonder if this shouldn't loop indefinitely or always log until we successfully connect, by logging every second we risk rotating and potentially losing logs if for some reason the device never makes a connection.

Reply:
Updated the logic to print every 10secs with total secs taken for retry and the port